docs/posts/llms-and-data/index.qmd

```diff
@@ -0,0 +1,258 @@
+---
+title: LLMs and data
+author: "Cody Peterson"
+date: "2023-09-27"
+freeze: auto
+categories:
+    - blog
+    - ai
+    - llm
+---
+
+## Introduction
+
+While large-language models (LLMs) have been around for years, recent versions
+and innovations have made it possible to create instruction-following,
+conversational bots that can perform tasks on behalf of the user. The
+thought of using natural language to transform and analyze data is appealing.
+
+## Approaches
+
+When discussed internally at Voltron Data, we identified three distinct
+approaches to applying LLMs to data analytics that can be implemented today:
+
+1. LLM writes SQL
+2. LLM writes a subroutine (Python UDF or another language)
+3. Use LLM in a subroutine
+
+While these three approaches are not an exhaustive list of how LLMs can be
+applied to data, they can be easily understood and implemented with
+Ibis and [Marvin](https://www.askmarvin.ai/welcome/what_is_marvin/), an AI
+engineering framework for building natural language interfaces. Together with
+these two open-source tools, we can build a natural language interface for
+data analytics that supports 18+ backends. But first, let's demonstrate the
+three approaches in code.
+
+### Approach 1: LLM writes SQL
+
+State of the art (SoTA) LLMs are decent at generating SQL out of the box. We can
+be clever to handle errors, retries, and more, but in its simplest form:
+
+```{python}
+# | code-fold: true
+import ibis  # <1>
+import marvin  # <1>
+
+from dotenv import load_dotenv  # <1>
+
+load_dotenv()  # <2>
+
+con = ibis.connect("duckdb://penguins.ddb")  # <3>
+t = ibis.examples.penguins.fetch()  # <3>
+t = con.create_table("penguins", t.to_pyarrow(), overwrite=True)  # <3>
+```
+
+1. Import the libraries we need.
+2. Load the environment variable to setup Marvin to call our OpenAI account.
+3. Setup the demo datain an Ibis backend.
+
+```{python}
+import ibis  # <1>
+import marvin  # <1>
+
+from ibis.expr.schema import Schema  # <1>
+from ibis.expr.types.relations import Table  # <1>
+
+ibis.options.interactive = True  # <2>
+marvin.settings.llm_model = "openai/gpt-4"  # <2>
+
+con = ibis.connect("duckdb://penguins.ddb")  # <3>
+t = con.table("penguins")  # <3>
+```
+
+1. Import Ibis and Marvin.
```

Comment:
What role is Ibis actually playing here, besides submitting the query and pretty-printing the results? Couldn't this just use DBAPI and achieve the same thing? If not can you explain why not?

Reply:
- easily handoff to LLMs (notice using Ibis to get the schema, table name, nice error messages to feed back, etc)
- translation across SQL dialects (you could just do this with SQLglot)
- access to Ibis for mixed Python/natural language analysis (use language OR Ibis code, mix and match)
- 18+ data platforms, thus this can be a universal standard for AI + data 
- once working reliably, write Ibis code that (somewhat) avoids the SQL dialect issues and works on 18+ backends
- then way down the line, Substrait via Ibis to really avoid those issues

will add some more explanation here